Validator panics when adding keys to remote signer file

### Describe the bug

The validator panics when adding keys to remote signer file.

### Has this worked before in a previous version?

_No response_

### 🔬 Minimal Reproduction

1. Start with remote signer file
2. Add validator keys to  external_keys.txt
3. It throws this sometimes

### Error

```text
time="2024-11-11 23:37:23" level=info msg="Successfully initialized file watcher" path="/data/external_keys.txt" prefix=remote-keymanager
time="2024-11-11 23:37:23" level=info msg="Remote signer key file updated" prefix=remote-keymanager
time="2024-11-11 23:37:23" level=info msg="Remote signer key file updated" prefix=remote-keymanager
panic: runtime error: index out of range [0] with length 0

goroutine 52 [running]:
github.com/prysmaticlabs/prysm/v5/validator/client/beacon-api.(*beaconApiValidatorClient).validatorsStatusResponse(0xc00043c000, {0x1ae6c70, 0xc0006914a0}, {0x4410340, 0x0, 0xc000508001?}, {0x4410340, 0x0, 0x0})
        validator/client/beacon-api/status.go:104 +0xe6d
github.com/prysmaticlabs/prysm/v5/validator/client/beacon-api.(*beaconApiValidatorClient).multipleValidatorStatus(0xc00043c000, {0x1ae6c70, 0xc0006914a0}, 0xc000753260)
        validator/client/beacon-api/status.go:35 +0xbf
github.com/prysmaticlabs/prysm/v5/validator/client/beacon-api.(*beaconApiValidatorClient).MultipleValidatorStatus.func1()
        validator/client/beacon-api/beacon_api_validator_client.go:135 +0x25
github.com/prysmaticlabs/prysm/v5/validator/client/beacon-api.wrapInMetrics[...]({0x1523119?, 0x17}, 0xc0005dadc0)
        validator/client/beacon-api/beacon_api_validator_client.go:312 +0x53
github.com/prysmaticlabs/prysm/v5/validator/client/beacon-api.(*beaconApiValidatorClient).MultipleValidatorStatus(0xc00043c000, {0x1ae6c70?, 0xc0006914a0?}, 0xc000753260)
        validator/client/beacon-api/beacon_api_validator_client.go:134 +0x11b
github.com/prysmaticlabs/prysm/v5/validator/client.(*validator).updateValidatorStatusCache(0xc0001b0fc8, {0x1ae6c70, 0xc0006914a0}, {0x4410340?, 0x0?, 0xc0dbc327b0?})
        validator/client/validator.go:1250 +0x209
github.com/prysmaticlabs/prysm/v5/validator/client.(*validator).HandleKeyReload(0xc0001b0fc8, {0x1ae6c70?, 0xc0006914a0?}, {0x4410340, 0x0, 0x0})
        validator/client/key_reload.go:19 +0x108
github.com/prysmaticlabs/prysm/v5/validator/client.onAccountsChanged({0x1ae6c70, 0xc0006914a0}, {0x1b0c130, 0xc0001b0fc8}, {0x4410340?, 0x2d1801?, 0xc0dfeee9c0?}, 0xc000988960)
        validator/client/runner.go:140 +0x6b
github.com/prysmaticlabs/prysm/v5/validator/client.run({0x1ae6c70, 0xc0006914a0}, {0x1b0c130, 0xc0001b0fc8})
        validator/client/runner.go:134 +0x1265
created by github.com/prysmaticlabs/prysm/v5/validator/client.(*ValidatorService).Start in goroutine 316
        validator/client/service.go:221 +0xd0b
```


### Platform(s)

Linux (x86)

### What version of Prysm are you running? (Which release)

5.1.2 current docker stable

### Anything else relevant (validator index / public key)?

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -78,6 +78,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 - Fix keymanager API so that get keys returns an empty response instead of a 500 error when using an unsupported keystore.
 - Small log imporvement, removing some redundant or duplicate logs
 - EIP7521 - Fixes withdrawal bug by accounting for pending partial withdrawals and deducting already withdrawn amounts from the sweep balance. [PR](https://github.com/prysmaticlabs/prysm/pull/14578)
+- Fix panic in validator REST mode when updating status to all keys removed
 
 
 ### Security

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -79,7 +79,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 - Small log imporvement, removing some redundant or duplicate logs
 - EIP7521 - Fixes withdrawal bug by accounting for pending partial withdrawals and deducting already withdrawn amounts from the sweep balance. [PR](https://github.com/prysmaticlabs/prysm/pull/14578)
 - unskip electra merkle spec test
-- Fix panic in validator REST mode when updating status to all keys removed
+- Fix panic in validator REST mode when checking status after removing all keys
 
 ### Security
 

--- a/beacon-chain/rpc/prysm/v1alpha1/validator/status.go
+++ b/beacon-chain/rpc/prysm/v1alpha1/validator/status.go
@@ -287,6 +287,9 @@ func (vs *Server) validatorStatus(
 		Status:          ethpb.ValidatorStatus_UNKNOWN_STATUS,
 		ActivationEpoch: params.BeaconConfig().FarFutureEpoch,
 	}
+	if len(pubKey) == 0 {
+		return resp, nonExistentIndex
+	}
 	vStatus, idx, err := statusForPubKey(headState, pubKey)
 	if err != nil && !errors.Is(err, errPubkeyDoesNotExist) {
 		tracing.AnnotateError(span, err)

--- a/validator/client/beacon-api/status.go
+++ b/validator/client/beacon-api/status.go
@@ -52,7 +52,7 @@ func (c *beaconApiValidatorClient) validatorsStatusResponse(ctx context.Context,
 ) {
 	// if no parameters are provided we should just return an empty response
 	if len(inPubKeys) == 0 && len(inIndexes) == 0 {
-		return make([][]byte, 0), make([]primitives.ValidatorIndex, 0), make([]*ethpb.ValidatorStatusResponse, 0), nil
+		return [][]byte{}, []primitives.ValidatorIndex{}, []*ethpb.ValidatorStatusResponse{}, nil
 	}
 	// Represents the target set of keys
 	stringTargetPubKeysToPubKeys := make(map[string][]byte, len(inPubKeys))

--- a/validator/client/beacon-api/status.go
+++ b/validator/client/beacon-api/status.go
@@ -50,6 +50,10 @@ func (c *beaconApiValidatorClient) validatorsStatusResponse(ctx context.Context,
 	[]*ethpb.ValidatorStatusResponse,
 	error,
 ) {
+	// if no parameters are provided we should just return an empty response
+	if len(inPubKeys) == 0 && len(inIndexes) == 0 {
+		return make([][]byte, 0), make([]primitives.ValidatorIndex, 0), make([]*ethpb.ValidatorStatusResponse, 0), nil
+	}
 	// Represents the target set of keys
 	stringTargetPubKeysToPubKeys := make(map[string][]byte, len(inPubKeys))
 	stringTargetPubKeys := make([]string, len(inPubKeys))
@@ -78,6 +82,7 @@ func (c *beaconApiValidatorClient) validatorsStatusResponse(ctx context.Context,
 		return nil, nil, nil, errors.Wrap(err, "failed to get state validators")
 	}
 
+	// TODO: we should remove this API call
 	validatorsCountResponse, err := c.prysmChainClient.ValidatorCount(ctx, "head", nil)
 	if err != nil && !errors.Is(err, iface.ErrNotSupported) {
 		return nil, nil, nil, errors.Wrap(err, "failed to get total validator count")

--- a/validator/client/beacon-api/status_test.go
+++ b/validator/client/beacon-api/status_test.go
@@ -215,32 +215,23 @@ func TestMultipleValidatorStatus_Nominal(t *testing.T) {
 	assert.DeepEqual(t, &expectedValidatorStatusResponse, actualValidatorStatusResponse)
 }
 
-func TestMultipleValidatorStatus_Error(t *testing.T) {
+func TestMultipleValidatorStatus_No_Keys(t *testing.T) {
 	ctrl := gomock.NewController(t)
 	defer ctrl.Finish()
 
 	ctx := context.Background()
 	stateValidatorsProvider := mock.NewMockStateValidatorsProvider(ctrl)
 
-	stateValidatorsProvider.EXPECT().StateValidators(
-		gomock.Any(),
-		gomock.Any(),
-		[]primitives.ValidatorIndex{},
-		nil,
-	).Return(
-		&structs.GetValidatorsResponse{},
-		errors.New("a specific error"),
-	).Times(1)
-
 	validatorClient := beaconApiValidatorClient{stateValidatorsProvider: stateValidatorsProvider}
 
-	_, err := validatorClient.MultipleValidatorStatus(
+	resp, err := validatorClient.MultipleValidatorStatus(
 		ctx,
 		&ethpb.MultipleValidatorStatusRequest{
 			PublicKeys: [][]byte{},
 		},
 	)
-	require.ErrorContains(t, "failed to get validators status response", err)
+	require.NoError(t, err)
+	require.DeepEqual(t, &ethpb.MultipleValidatorStatusResponse{}, resp)
 }
 
 func TestGetValidatorsStatusResponse_Nominal_SomeActiveValidators(t *testing.T) {

--- a/validator/client/validator.go
+++ b/validator/client/validator.go
@@ -1224,6 +1224,10 @@ func (v *validator) filterAndCacheActiveKeys(ctx context.Context, pubkeys [][fie
 
 // updateValidatorStatusCache updates the validator statuses cache, a map of keys currently used by the validator client
 func (v *validator) updateValidatorStatusCache(ctx context.Context, pubkeys [][fieldparams.BLSPubkeyLength]byte) error {
+	if len(pubkeys) == 0 {
+		v.pubkeyToStatus = make(map[[fieldparams.BLSPubkeyLength]byte]*validatorStatus, 0)
+		return nil
+	}
 	statusRequestKeys := make([][]byte, 0)
 	for _, k := range pubkeys {
 		statusRequestKeys = append(statusRequestKeys, k[:])

--- a/validator/client/validator_test.go
+++ b/validator/client/validator_test.go
@@ -2899,4 +2899,9 @@ func TestUpdateValidatorStatusCache(t *testing.T) {
 		require.Equal(t, mockResponse.Statuses[i], status.status)
 		require.Equal(t, mockResponse.Indices[i], status.index)
 	}
+
+	err = v.updateValidatorStatusCache(ctx, nil)
+	assert.NoError(t, err)
+	// make sure the value is 0
+	assert.Equal(t, 0, len(v.pubkeyToStatus))
 }

